File: String/242-Valid-anagram.java
// 242. Valid Anagram
// Easy
// Topics
// Companies
// Given two strings s and t, return true if t is an anagram of s, and false
// otherwise.

// An Anagram is a word or phrase formed by rearranging the letters of a
// different word or phrase, typically using all the original letters exactly
// once.

// Example 1:

// Input: s = "anagram", t = "nagaram"
// Output: true
// Example 2:

// Input: s = "rat", t = "car"
// Output: false

// Constraints:

// 1 <= s.length, t.length <= 5 * 104
// s and t consist of lowercase English letters.
import java.util.HashMap;

class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        HashMap<Character, Integer> countS = new HashMap<>();
        HashMap<Character, Integer> countT = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            countS.put(s.charAt(i), countS.getOrDefault(s.charAt(i), 0) + 1);
            countT.put(t.charAt(i), countT.getOrDefault(t.charAt(i), 0) + 1);
        }

        if (!countS.keySet().equals(countT.keySet())) {
            return false;
        }

        for (char c : countS.keySet()) {
            if (!countS.get(c).equals(countT.get(c))) {
                return false;
            }
        }

        return true;
    }
}

File: String/242-Valid-anagram.ts
function isAnagram(s: string, t: string): boolean {
    if (s.length != t.length) {
        return false;
    }

    let countS: Map<string, number> = new Map();
    let countT: Map<string, number> = new Map();

    for (let i = 0; i < s.length; i++) {
        countS.set(s[i], (countS.get(s[i]) || 0) + 1);
        countT.set(t[i], (countT.get(t[i]) || 0) + 1);
    }

    if (countS.keys !== countT.keys) {
        return false;
    }

    for (let key of countS.keys()) {
        if (countS.get(key) !== countT.get(key)) {
            return false;
        }
    }
    return true;
};